File: app/demo/gh.py
# Write a function to get the latest commits from a GitHub repository
# using the GitHub API.
# write the committer name and commit message into an excel file in one sheet

import requests
import pandas as pd
from datetime import datetime


def get_latest_commits(repo_owner, repo_name, token, num_commits=10):
    """
    Fetches the latest commits from a specified GitHub repository.
    Args:
        repo_owner (str): The owner of the repository (username or organization name).
        repo_name (str): The name of the repository.
        token (str): A personal access token for authenticating with the GitHub API.
        num_commits (int, optional): The number of commits to retrieve. Defaults to 10.
    Returns:
        list[dict] or None: A list of dictionaries containing commit data with the keys:
            - 'Committer Name': The name of the committer.
            - 'Commit Message': The commit message.
            Returns None if the request fails.
    Raises:
        None: This function handles HTTP errors internally and prints the error status code.
    """
    url = f"https://api.github.com/repos/{repo_owner}/{repo_name}/commits"ṇ
    params = {
        "per_page": num_commits,
        "page": 1
    }
    headers = { 
        "Authorization": f"token {token}"
    }
    response = requests.get(url, params=params, headers=headers)
    
    if response.status_code == 200:
        commits = response.json()
        commit_data = []
        
        for commit in commits:
            committer_name = commit['commit']['committer']['name']
            commit_message = commit['commit']['message']
            commit_data.append({
                'Committer Name': committer_name,
                'Commit Message': commit_message
            })
        
        return commit_data
    else:
        print(f"Error: {response.status_code}")
        return None

def write_to_excel(commit_data, file_name=None):
    if file_name is None:
        timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
        file_name = f'latest_commits_{timestamp}.xlsx'
    df = pd.DataFrame(commit_data)
    df.to_excel(file_name, index=False)
    print(f"Data written to {file_name}")

File: app/demo/gh.cs
// Write a function to get the latest commits from a GitHub repository
// using the GitHub API.
// write the committer name, commit message and the date into an excel file in one sheet

using System;
using System.Collections.Generic;
using System.IO;
using OfficeOpenXml;
using System.Net.Http;
using System.Net.Http.Headers;

namespace GitHubCommitsToExcel
{
    class Program
    {
        static void Main(string[] args)
        {
            string repoOwner = "octocat"; // Replace with the repository owner
            string repoName = "Hello-World"; // Replace with the repository name
            string excelFilePath = "GitHubCommits.xlsx";

            var commits = GetLatestCommits(repoOwner, repoName).Result;
            WriteCommitsToExcel(commits, excelFilePath);
            Console.WriteLine($"Commits written to {excelFilePath}");
        }

        public class CommitInfo
        {
            public string CommitterName { get; set; }
            public string CommitMessage { get; set; }
            public DateTime CommitDate { get; set; }
        }

        static async System.Threading.Tasks.Task<List<CommitInfo>> GetLatestCommits(string owner, string repo)
        {
            var commits = new List<CommitInfo>();
            using (var client = new HttpClient())
            {
                client.BaseAddress = new Uri("https://api.github.com/");
                client.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("Mozilla", "5.0"));
                client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/vnd.github.v3+json"));

                var response = await client.GetAsync($"/repos/{owner}/{repo}/commits");
                response.EnsureSuccessStatusCode();

                var commitData = await response.Content.ReadAsAsync<dynamic[]>();

                foreach (var item in commitData)
                {
                    try
                    {
                        commits.Add(new CommitInfo
                        {
                            CommitterName = item?.commit?.committer?.name ?? "Unknown",
                            CommitMessage = item?.commit?.message ?? "No message",
                            CommitDate = item?.commit?.committer?.date != null 
                                ? DateTime.Parse(item.commit.committer.date.ToString()) 
                                : DateTime.MinValue
                        });
                    }
                    catch (Exception ex)
                    {
                        // Log the error and continue with default values
                        Console.WriteLine($"Error processing commit: {ex.Message}");
                        commits.Add(new CommitInfo
                        {
                            CommitterName = "Error",
                            CommitMessage = "Failed to parse commit data",
                            CommitDate = DateTime.MinValue
                        });
                    }
                }
            }
            return commits;
        }

        static void WriteCommitsToExcel(List<CommitInfo> commits, string filePath)
        {
            ExcelPackage.LicenseContext = LicenseContext.NonCommercial;
            using (var package = new ExcelPackage())
            {
                var worksheet = package.Workbook.Worksheets.Add("Commits");
                worksheet.Cells[1, 1].Value = "Committer Name";
                worksheet.Cells[1, 2].Value = "Commit Message";
                worksheet.Cells[1, 3].Value = "Commit Date";

                for (int i = 0; i < commits.Count; i++)
                {
                    worksheet.Cells[i + 2, 1].Value = commits[i].CommitterName;
                    worksheet.Cells[i + 2, 2].Value = commits[i].CommitMessage;
                    worksheet.Cells[i + 2, 3].Value = commits[i].CommitDate.ToString("yyyy-MM-dd HH:mm:ss");
                }
                package.SaveAs(new FileInfo(filePath));
            }
        }
    }
}